client/main.cpp

```diff
@@ -61,51 +53,57 @@ int main(int argc, char **argv) {
     load_root_certificates(ctx);
     ctx.set_verify_mode(ssl::verify_peer);
 
-    boost::asio::io_context ioc;
-    nibaclient::client_session client_session(host, port, ioc, ctx);
-    std::mutex processor_mutex;
-    std::condition_variable processor_cv;
-    bool processed = true;
-
-    std::thread([&ioc, &instream, &client_session, &processor_mutex, &processor_cv, &processed] {
-        // separate io thread so that getline doesn't block our websocket pingpong
-        std::string line;
-        while (std::getline(instream, line)) {
-            if (line.empty())
-                continue;            
-            if (line == "exit" || line == "quit")
-                break;
-            std::unique_lock<std::mutex> lock(processor_mutex);
-            processor_cv.wait(lock, [&processed] { return processed; });
-            auto earliest = client_session.earliest();
-            auto now = std::chrono::high_resolution_clock::now();
-            if (earliest > now) {
-                // if we do auto its nano seconds, fine
-                auto delay = earliest - now;
-                std::cout << "cooldown " << delay.count() << "ns" << std::endl;
-                std::this_thread::sleep_for(delay);
-            }
-            processed = false;
-            try {
-                // post the work to ioc
-                // we do not do any network io here
-                ioc.post([&client_session, &processor_mutex, &processor_cv, &processed, line]() {
-                    std::unique_lock<std::mutex> lock(processor_mutex);
-                    client_session.handle_cmd(line);
-                    processed = true;
-                    lock.unlock();
-                    processor_cv.notify_one();
-                });
-            } catch (std::exception &e) {
-                std::cout << e.what() << std::endl;
-                break;
-            }
+    boost::asio::io_context ioc{};
+    auto work_guard = boost::asio::make_work_guard(ioc);
```

Comment:
What's the use of work_guard?

Reply:
So the `io_context` doesn't run out of work - eg. when we spawn the thread to `ioc.run()` it will be blocked (without the work guard it will just return). We didn't need this before because we used to spawn a coroutine (or have other work running) before calling `ioc.run()`